src/main/java/com/silzila/querybuilder/RelativeCalendarFilter.java

```diff
@@ -0,0 +1,70 @@
+package com.silzila.querybuilder;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.h2.expression.Alias;
+
+import com.silzila.helper.AilasMaker;
+import com.silzila.payload.request.Dimension;
+import com.silzila.payload.request.Measure;
+import com.silzila.payload.request.Query;
+
+public class RelativeCalendarFilter {
+
+    public static String buildRelativeCalendarFilter(Query req, String CTEquery, String relativeType) {
+
+        Map<String, Integer> aliasNumbering = new HashMap<>();
+
+        String CTEqueryString = "WITH CTE AS (\n" + CTEquery + "\n) \n";
+
+        List<Dimension> dimensions = req.getDimensions();
+
+        List<Measure> measures = req.getMeasures();
+
+        List<String> selectWindowClause = new ArrayList<>();
+
+        for (Measure measure : measures) {
+
+            String alias = AilasMaker.aliasing(measure.getFieldName(), aliasNumbering);
+            String field = measure.getAggr().name() + "(" + alias
+                    + ") over (" + partitionBy(dimensions) + ")";
+            selectWindowClause.add(field);
+        }
+
+        StringBuilder finalRelativeQuery = new StringBuilder();
+        finalRelativeQuery.append(CTEqueryString);
+        finalRelativeQuery.append("Select * , ");
+        finalRelativeQuery.append(String.join(", ", selectWindowClause));
+        finalRelativeQuery.append("\n from CTE");
+
+        return finalRelativeQuery.toString();
+    }
+
+    public static String partitionBy(List<Dimension> dimensions) {
+
+        String partitionQuery = "";
+        List<Dimension> DimensionWithoutDate = new ArrayList<>();
+        for (int i = 0; i < dimensions.size(); i++) {
+            Dimension dim = dimensions.get(i);
+            if (!("date".equals(dim.getDataType().toString()) || "timestamp".equals(dim.getDataType().toString()))) {
```

Comment:
I would write it little differently, but it is not a big deal. You don't need to change but just keep in mind that try to avoid computing the same thing again and again if possible.

String dimType = dim.getDataType().toString()
if ( !( "date".equals(dimType) || "timestamp".equals(dimType)))